server/src/main/java/org/opensearch/index/remote/RemoteRefreshSegmentPressureService.java

```diff
@@ -0,0 +1,304 @@
+     *
+     * @return true if enabled, else false.
+     */
+    public boolean isSegmentsUploadBackpressureEnabled() {
+        return pressureSettings.isRemoteRefreshSegmentPressureEnabled();
+    }
+
+    public void validateSegmentsUploadLag(ShardId shardId) {
+        RemoteRefreshSegmentTracker remoteRefreshSegmentTracker = getRemoteRefreshSegmentTracker(shardId);
+        // Check if refresh checkpoint (a.k.a. seq number) lag is 2 or below - this is to handle segment merges that can
+        // increase the bytes to upload almost suddenly.
+        if (remoteRefreshSegmentTracker.getSeqNoLag() <= 1) {
+            return;
+        }
+
+        for (LagValidator lagValidator : lagValidators) {
+            if (lagValidator.validate(remoteRefreshSegmentTracker, shardId) == false) {
+                remoteRefreshSegmentTracker.incrementRejectionCount(lagValidator.name());
+                throw new OpenSearchRejectedExecutionException(lagValidator.rejectionMessage(remoteRefreshSegmentTracker, shardId));
+            }
+        }
+    }
+
+    void updateUploadBytesMovingAverageWindowSize(int updatedSize) {
+        updateMovingAverageWindowSize(RemoteRefreshSegmentTracker::updateUploadBytesMovingAverageWindowSize, updatedSize);
+    }
+
+    void updateUploadBytesPerSecMovingAverageWindowSize(int updatedSize) {
+        updateMovingAverageWindowSize(RemoteRefreshSegmentTracker::updateUploadBytesPerSecMovingAverageWindowSize, updatedSize);
+    }
+
+    void updateUploadTimeMsMovingAverageWindowSize(int updatedSize) {
+        updateMovingAverageWindowSize(RemoteRefreshSegmentTracker::updateUploadTimeMsMovingAverageWindowSize, updatedSize);
+    }
+
+    void updateMovingAverageWindowSize(BiConsumer<RemoteRefreshSegmentTracker, Integer> biConsumer, int updatedSize) {
+        trackerMap.values().forEach(tracker -> biConsumer.accept(tracker, updatedSize));
+    }
+
+    /**
+     * Abstract class for validating if lag is acceptable or not.
+     *
+     * @opensearch.internal
+     */
+    private static abstract class LagValidator {
+
+        final RemoteRefreshSegmentPressureSettings pressureSettings;
+
+        private LagValidator(RemoteRefreshSegmentPressureSettings pressureSettings) {
+            this.pressureSettings = pressureSettings;
+        }
+
+        /**
+         * Validates the lag and returns value accordingly.
+         *
+         * @param pressureTracker tracker which holds information about the shard.
+         * @param shardId         shard id of the {@code IndexShard} currently being validated.
+         * @return true if successfully validated that lag is acceptable.
+         */
+        abstract boolean validate(RemoteRefreshSegmentTracker pressureTracker, ShardId shardId);
+
+        /**
+         * Returns the name of the lag validator.
+         *
+         * @return the name using class name.
+         */
+        final String name() {
+            return this.getClass().getSimpleName();
+        }
```

Comment:
instead of class name , lets enforce the classes to define this function to make it more readable .